File: backend/app/schemas/quiz.py
from pydantic import BaseModel, Field # type: ignore
from typing import List, Dict, Optional

from app.schemas.mcq import MCQItem


class QuizCreateRequest(BaseModel):
    query: str
    original_filename: str
    mcqs: List[MCQItem]


class QuizCreateResponse(BaseModel):
    message: str
    session_id: str
    query: str
    original_filename: str
    total_questions: int
    mcqs: List[MCQItem]


class QuizSubmitRequest(BaseModel):
    session_id: str
    answers: Dict[str, str] = Field(
        ...,
        description="Map of question_id to selected option label, e.g. {'1': 'B', '2': 'A'}"
    )


class QuestionReview(BaseModel):
    question_id: str
    question: str
    user_answer: Optional[str]
    correct_answer: str
    is_correct: bool
    difficulty: str
    explanation: str
    confidence_score: float
    is_verified: bool


class DifficultyStats(BaseModel):
    total: int
    correct: int
    accuracy: float


class QuizSubmitResponse(BaseModel):
    message: str
    session_id: str
    total_questions: int
    correct_count: int
    incorrect_count: int
    percentage: float
    difficulty_breakdown: Dict[str, DifficultyStats]
    question_reviews: List[QuestionReview]
    weak_areas: List[str]

File: ai-mcq-frontend/src/types/quiz.ts
export type DifficultyLevel = "easy" | "medium" | "hard" | string;

export interface QuizOption {
  label: string;
  text: string;
}

export interface QuizQuestion {
  id?: string;
  question_id?: string;
  question: string;
  options: QuizOption[];
  correct_answer: string; // backend returns label like "A", "B"
  difficulty: DifficultyLevel;
  explanation?: string;
  confidence_score?: number;
  is_verified?: boolean;
  verification_notes?: string[];
}

export interface GenerateQuizResponse {
  message: string;
  original_filename: string;
  saved_path: string;
  num_pages: number;
  text_length: number;
  topic?: string;
  query: string;
  chunks_created: number;
  keyphrases: Array<{
    keyword: string;
    score: number;
  }>;
  focused_context: string;
  mcqs: QuizQuestion[];
}

export interface QuizCreateRequest {
  query: string;
  original_filename: string;
  mcqs: QuizQuestion[];
}

export interface QuizCreateResponse {
  message: string;
  session_id: string;
  query: string;
  original_filename: string;
  total_questions: number;
  mcqs: QuizQuestion[];
}

export interface SubmitQuizRequest {
  session_id: string;
  answers: Record<string, string>; // question_id -> selected option label
}

export interface DifficultyStats {
  total: number;
  correct: number;
  accuracy?: number;
}

export interface QuestionReview {
  question_id: string;
  question: string;
  user_answer: string;
  correct_answer: string;
  is_correct: boolean;
  difficulty: string;
  explanation?: string;
  confidence_score?: number;
  is_verified?: boolean;
}

export interface QuizResult {
  message: string;
  session_id: string;
  total_questions: number;
  correct_count: number;
  incorrect_count: number;
  percentage: number;
  difficulty_breakdown: Record<string, DifficultyStats>;
  question_reviews: QuestionReview[];
  weak_areas: string[];
}